metrics-operator/controllers/common/providers/dynatrace/dynatrace_dql.go

```diff
@@ -85,39 +96,91 @@ func NewKeptnDynatraceDQLProvider(k8sClient client.Client, opts ...KeptnDynatrac
 	return provider
 }
 
-// EvaluateQuery fetches the SLI values from dynatrace provider
 func (d *keptnDynatraceDQLProvider) EvaluateQuery(ctx context.Context, metric metricsapi.KeptnMetric, provider metricsapi.KeptnMetricsProvider) (string, []byte, error) {
 	if err := d.ensureDTClientIsSetUp(ctx, provider); err != nil {
 		return "", nil, err
 	}
-	// submit DQL
-	dqlHandler, err := d.postDQL(ctx, metric.Spec.Query)
+
+	b, status, err := d.postDQL(ctx, metric)
 	if err != nil {
 		d.log.Error(err, "Error while posting the DQL query", "query", metric.Spec.Query)
 		return "", nil, err
 	}
-	// attend result
-	results, err := d.getDQL(ctx, *dqlHandler)
+
+	results, err := d.parseDQLResults(b, status)
 	if err != nil {
-		d.log.Error(err, "Error while waiting for DQL query", "query", dqlHandler)
 		return "", nil, err
 	}
 
-	// parse result
 	if len(results.Records) > 1 {
 		d.log.Info("More than a single result, the first one will be used")
 	}
-	if len(results.Records) == 0 {
-		return "", nil, ErrInvalidResult
-	}
+
 	r := fmt.Sprintf("%f", results.Records[0].Value.Avg)
-	b, err := json.Marshal(results)
+	b, err = json.Marshal(results)
+	if err != nil {
+		d.log.Error(err, "Error marshaling DQL results")
+	}
+
+	return r, b, nil
+}
+
+func (d *keptnDynatraceDQLProvider) EvaluateQueryForStep(ctx context.Context, metric metricsapi.KeptnMetric, provider metricsapi.KeptnMetricsProvider) ([]string, []byte, error) {
+	if err := d.ensureDTClientIsSetUp(ctx, provider); err != nil {
+		return nil, nil, err
+	}
+
+	b, status, err := d.postDQL(ctx, metric)
+	if err != nil {
+		d.log.Error(err, "Error while posting the DQL query", "query", metric.Spec.Query)
+		return nil, nil, err
+	}
+
+	results, err := d.parseDQLResults(b, status)
+	if err != nil {
+		return nil, nil, err
+	}
```

Comment:
maybe we can think of encapsulating these lines into a new function since they are the same across both invokations

Reply:
Done